message/messaging_service.cc

```diff
@@ -284,6 +284,32 @@ bool messaging_service::is_same_rack(inet_address addr) const {
     return topo.get_rack(addr) == topo.get_rack();
 }
 
+future<> messaging_service::ban_host(locator::host_id id) {
```

Comment:
IIRC in the last review I asked to drop client as well. This will be used for node isolation as well.

Reply:
And I answered that I cannot do this in this PR because I'm blocked by https://github.com/scylladb/scylladb/issues/6403
In this comment https://github.com/scylladb/scylladb/pull/13850#issuecomment-1548079128:

> I didn't implement dropping/banning of outgoing connections, because it's blocked by https://github.com/scylladb/scylladb/issues/6403. When we send a message, in general, we don't have access to the host_id of the destination inside messaging_service.
AFAIK @bhalevy is working on extending msg_addr with host_ids. Then it will be possible.

> Still, I think that banning incoming connections provides 99%-100% of the value that this PR is supposed to provide: it effectively isolates removed nodes, as the added test illustrates.

> I propose that in the scope of this PR, we solve banning incoming connections, and leave banning outgoing connections to a follow-up, after we have https://github.com/scylladb/scylladb/issues/6403.

Reply:
Even if I could to this (which I don't see how without #6403), it's IMO better to develop this in a bunch of smaller PRs instead of one giant PR that does everything.

Reply:
I missed your answer (PRs are crap), but I do not understand why can't we map ids to ips as interim solution?
If we do have a mapping adding the code that drops client is easy. IIRC we already have a function that does it. But OK. Lets leave it for the very near future (we will need it for `ignore-nodes` option.

Reply:
> I missed your answer (PRs are crap), but I do not understand why can't we map ids to ips as interim solution?

It would not work very well in presence of changing IPs or losing IP mappings - for example
1. we ban host A which is mapped to IP X
2. at some point another host B reuses IP X
3. we get message from IP X, our mapping says it's host A which is banned

or, when we remove host A, we eventually stop gossiping its IP and lose its mapping, so we'd also lose the ban

I think such interim solution would bring more problems than it solves issues, we really need to fix #6403, definitely before we GA topology.

Reply:
#6403 will not magically use UUIDs on the wire.  The scenario you described is exactly what can happen with this patch, but since you do not use ip to check what is banned, but the UUID that you get as part of CLIEN_ID the code works as indented though #6403 is not solved. What is missing is dropping clients as well since you cannot map directly from UUID to a client. The worst that can happen (and I am not sure it can) you will map ID to the wrong IP an drop incorrect client. New one will be created (though some RPCs may fail why they should not have).

Reply:
> The scenario you described is exactly what can happen with this patch, but since you do not use ip to check what is banned, but the UUID that you get as part of CLIEN_ID the code works as indented

So you're saying that the scenario can happen with this patch, but actually it cannot? :)

To drop clients today I would have to map from IPs to UUIDs, not the other way. And this suffers from the problems I described.
The correct solution is to identify clients by UUIDs in the first place. We don't even need UUIDs for the write for that. We just need to know who we're sending the message to.
Which we will be able to do after #6403:
>  AFAIK @bhalevy is working on extending msg_addr with host_ids. Then it will be possible.

Reply:
> > The scenario you described is exactly what can happen with this patch, but since you do not use ip to check what is banned, but the UUID that you get as part of CLIEN_ID the code works as indented
> 
> So you're saying that the scenario can happen with this patch, but actually it cannot? :)

I am saying it can, but it does not produce the effect you claim because we do not ban by an IP, but by an ID.

> To drop clients today I would have to map from IPs to UUIDs, not the other way.

And the worst that may happen (as I described in my reply) you will drop incorrect client and it will be re-created.
And I am not even sure it may eventually happen because new node will not be able to claim old IP until we complete previous operation which includes banning. May be possible with replace op.

> And this suffers from the problems I described.

No. It does not because we ban by ID, not IP. We only drop clients by IP.

Reply:
So you suggest we only drop clients, but don't ban outgoing messages? What's the value of that, if the client will be recreated automatically on the next message?

Reply:
Right, for banning outgoing message we need to translate IP to ID on client creation or at least have @bhalevy's patch. But is it a problem? The cluster should be able to learn new mapping through gossiper. We allow incoming connections.

Reply:
IIUC if we get a gossip SYN, we don't update our metadata yet, we need to send back ACK and then receive ACK2 to update the metadata. So we also need to allow the outgoing ACK.

But maybe we could update the mapping earlier - immediately when we receive CLIENT_ID verb.
This would solve the potential availability problem.

But there's also the problem that we're not keeping mappings of nodes that we removed forever - they are TTLd ~1 hour. So after ~1 hour of banning a node, we would unban outgoing messages because we would no longer be able to map the IP to ID and notice that the ID is banned.

Reply:
And we cannot drop message if mapping is not available since a new node will not be able to communicate with a cluster. It becomes complicated but it does not look like https://github.com/scylladb/scylladb/issues/6403 is planned for the near future.

Reply:
#6403 is blocking raft topology from becoming GA, so we'll be able to do the banning properly before becoming GA as well.